erigon-lib/state/domain.go

```diff
@@ -1158,21 +1165,75 @@ func (d *Domain) integrateDirtyFiles(sf StaticFiles, txNumFrom, txNumTo uint64)
 
 // unwind is similar to prune but the difference is that it restores domain values from the history as of txFrom
 // context Flush should be managed by caller.
-func (dt *DomainRoTx) Unwind(ctx context.Context, rwTx kv.RwTx, step, txNumUnwindTo uint64) error {
+func (dt *DomainRoTx) Unwind(ctx context.Context, rwTx kv.RwTx, step, txNumUnwindTo uint64, diff *StateDiffDomain) error {
 	d := dt.d
 	//fmt.Printf("[domain][%s] unwinding domain to txNum=%d, step %d\n", d.filenameBase, txNumUnwindTo, step)
-	histRng, err := dt.ht.HistoryRange(int(txNumUnwindTo), -1, order.Asc, -1, rwTx)
-	if err != nil {
-		return fmt.Errorf("historyRange %s: %w", dt.ht.h.filenameBase, err)
-	}
+
 	sf := time.Now()
 	defer mxUnwindTook.ObserveDuration(sf)
 	mxRunningUnwind.Inc()
 	defer mxRunningUnwind.Dec()
+	restored := dt.NewWriter()
+	logEvery := time.NewTicker(time.Second * 30)
+	defer logEvery.Stop()
+
+	stepBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(stepBytes, ^step)
+	// Attempt to use the diff to unwind the domain
+	if diff != nil {
+		valsKV := diff.GetKeys()
+		keysCursor, err := rwTx.RwCursorDupSort(d.keysTable)
+		if err != nil {
+			return fmt.Errorf("create %s domain delete cursor: %w", d.filenameBase, err)
+		}
+		defer keysCursor.Close()
+
+		// First revert keys
+		for _, kv := range valsKV {
```

Comment:
`for _, kv := range valsKV {` does the copy elements - because it's non-pointer type. I think - it will speedup loop if rewrite it by:
```
for i := range valsKV {
   k, v := valsKV[i].Key, valsKV[i].Val
```